{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pan\n",
    "import numpy as np\n",
    "\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import ranksums\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### select strongest cis eqtl ### \n",
    "## run the findr 1_select_cis_eqtls.py## "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### prepare genotype data #############\n",
    "# genotype prepared by writing the 2_prepare_genotype_data.py file#### \n",
    "### the output is genotypes_binary_strongest_eqtl.csv file ########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Reorder the gene expression data matrices:\n",
      "#    Load data:\n",
      "#    Reordering:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3248: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider using pd.concat instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  if (await self.run_code(code, result,  async_=asy)):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       numbers_of_genes:  [2836, 5720, 2884]\n",
      "       genes with eQTLs:  2884\n",
      "       all genes:  5720\n",
      "    Check if all expected columns are present: True\n",
      "    number of columns selected: 5720\n"
     ]
    }
   ],
   "source": [
    "## reorder expression data ### \n",
    "## run the findr 3_reorder_expression_data.py ### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#yeastract= pan.read_csv(\"./../data/input/yeastract/DNABinding_AND_Expr_TF_act_OR_inh_RegulationMatrix_Documented_2020511.csv.gz\",header=0,index_col=0,\n",
    "                           # delimiter=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## read yeastract file ######\n",
    "yeastract= pan.read_csv(\"/data/users/cs20s037/CITNonLinear/findr/findrfiles/data/input/yeastract/DNABinding_AND_Expr_TF_act_OR_inh_RegulationMatrix_Documented_2020511.csv.gz\",header=0,index_col=0,\n",
    "                            delimiter=';')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11486"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(yeastract.sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### ensemble file downloaded and preprocessing done using the shell script file ##### \n",
    "## ensembl 109 used for yeast ## "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#  Selecting compatible subset of yeast regulatory network from yeastract with respect to elife paper.\n",
      "#  1. Load ensembl yeast gene library:\n"
     ]
    }
   ],
   "source": [
    "print('#  Selecting compatible subset of yeast regulatory network from yeastract with respect to elife paper.')\n",
    "print('#  1. Load ensembl yeast gene library:')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_full = pan.read_csv( '/data/users/cs20s037/CITNonLinear/findr/findrfiles/data/cleaned_genes_ensembl109_yeast_step1.txt', skiprows=0,\n",
    "                    names='chromosome position gene_str'.split(), \n",
    "                    usecols=[0,3,8],\n",
    "                    comment='#',\n",
    "                    sep='\\t| ', engine='python')\n",
    "\n",
    "df_full['id'] = df_full['gene_str'].str.split(':').str[1]\n",
    "full_gene_list = df_full['id'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#   The file used below only contains those entries with Name and ID.\n",
    "df = pan.read_csv(\"/data/users/cs20s037/CITNonLinear/findr/findrfiles/data/cleaned_genes_ensembl109_yeast_step2.txt\", skiprows=0,\n",
    "                    names='chromosome position gene_str name_str'.split(), \n",
    "                    usecols=[0,3,8,9],\n",
    "                    comment='#',\n",
    "                    sep='\\t| ', engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#  data frame as loaded\n",
      "     chromosome  position           gene_str   name_str\n",
      "0             I      1807    ID=gene:YAL068C  Name=PAU8\n",
      "1             I      7235    ID=gene:YAL067C  Name=SEO1\n",
      "2             I     13363  ID=gene:YAL064C-A  Name=TDA8\n",
      "3             I     24000    ID=gene:YAL063C  Name=FLO9\n",
      "4             I     31567    ID=gene:YAL062W  Name=GDH3\n",
      "...         ...       ...                ...        ...\n",
      "5155        XVI    924304    ID=gene:YPR194C  Name=OPT2\n",
      "5156        XVI    934034    ID=gene:YPR198W  Name=SGE1\n",
      "5157        XVI    938148    ID=gene:YPR199C  Name=ARR1\n",
      "5158        XVI    939279    ID=gene:YPR200C  Name=ARR2\n",
      "5159        XVI    939922    ID=gene:YPR201W  Name=ARR3\n",
      "\n",
      "[5160 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "print('#  data frame as loaded')\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#  data frame after column operations and sorting\n",
      "     chromosome  position           gene_str   name_str  name         id\n",
      "0             I      1807    ID=gene:YAL068C  Name=PAU8  PAU8    YAL068C\n",
      "1             I      7235    ID=gene:YAL067C  Name=SEO1  SEO1    YAL067C\n",
      "2             I     13363  ID=gene:YAL064C-A  Name=TDA8  TDA8  YAL064C-A\n",
      "3             I     24000    ID=gene:YAL063C  Name=FLO9  FLO9    YAL063C\n",
      "4             I     31567    ID=gene:YAL062W  Name=GDH3  GDH3    YAL062W\n",
      "...         ...       ...                ...        ...   ...        ...\n",
      "5155        XVI    924304    ID=gene:YPR194C  Name=OPT2  OPT2    YPR194C\n",
      "5156        XVI    934034    ID=gene:YPR198W  Name=SGE1  SGE1    YPR198W\n",
      "5157        XVI    938148    ID=gene:YPR199C  Name=ARR1  ARR1    YPR199C\n",
      "5158        XVI    939279    ID=gene:YPR200C  Name=ARR2  ARR2    YPR200C\n",
      "5159        XVI    939922    ID=gene:YPR201W  Name=ARR3  ARR3    YPR201W\n",
      "\n",
      "[5160 rows x 6 columns]\n",
      "#  number of genes: 5160\n"
     ]
    }
   ],
   "source": [
    "#   Create new column name and id\n",
    "df['name'] = df['name_str'].str.split('=').str[1]\n",
    "df['id'] = df['gene_str'].str.split(':').str[1]\n",
    "#   Sort it\n",
    "dfs=df.sort_values(axis=0,by='chromosome position'.split() )\n",
    "#   Make gene_name_dict\n",
    "print('#  data frame after column operations and sorting')\n",
    "print(dfs)\n",
    "df['name'] = df['name'].str.upper()\n",
    "df['id'] = df['id'].str.upper()\n",
    "dfi = dfs.set_index('name')\n",
    "gene_name_dict = dfi['id'].to_dict()\n",
    "print('#  number of genes:', len(gene_name_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def  relabel_matrix(matrix, index_name_dict, full_gene_list):\n",
    "    \"\"\" Function doc \n",
    "    \n",
    "        Use dictionnary to relabel rows and columns of matrix.\n",
    "        Intended to be used for instance with genes.\n",
    "        \n",
    "        Input:\n",
    "            matrix:     (regulatory) matrix with row and column labels of genes or transcription factors (TF). \n",
    "            index_name_dict:    dictionnary to map TF names to gene names (obtained from ensembl library).\n",
    "        \n",
    "        Output:\n",
    "            sub_matrix: a submatrix of the input matrix, that contains \n",
    "                        the rows and columns with labels\n",
    "                        that are compatible with the index_name_dict.\n",
    "    \"\"\"\n",
    "    old_cols = matrix.columns\n",
    "    old_rows = matrix.index.values\n",
    "    print('old_cols, n=',len(old_cols) )\n",
    "    print(old_cols)\n",
    "    print('old_rows, n=', len(old_rows))\n",
    "    print(old_rows)\n",
    "    print('#  parsing columns:')\n",
    "    col_subset, new_cols_dict = select_compatible_labels(old_cols, index_name_dict, full_gene_list)\n",
    "    print('#  parsing rows:')\n",
    "    row_subset, new_rows_dict = select_compatible_labels(old_rows, index_name_dict, full_gene_list)\n",
    "    sub_matrix = matrix.loc[ row_subset, col_subset ]\n",
    "    sub_matrix.rename( columns = new_cols_dict, inplace=True)\n",
    "    sub_matrix.rename( index = new_rows_dict, inplace=True)\n",
    "    print('  number of compatible columns',len(col_subset))\n",
    "    print('  number of compatible rows',len(row_subset))\n",
    "    return sub_matrix\n",
    "\n",
    "\n",
    "def  select_compatible_labels( old_labels, label_dict, full_label_list ):\n",
    "    \"\"\" Function doc \n",
    "    \n",
    "        Parse all old_labels to find those compatible with the label_dict.\n",
    "        Incompatible labels are printed in output for logging.\n",
    "        \n",
    "        Input: \n",
    "            old_labels:     labels of row or column.\n",
    "            label_dict:     dictionnary to map TF names to gene names (obtained from ensembl library).\n",
    "    \n",
    "        Output:\n",
    "            label_subset:   subset of compatible labels.\n",
    "            new_labels_dict:    dictionnary of compatible TF names to gene names.\n",
    "    \"\"\"\n",
    "    label_subset=[]\n",
    "    new_labels_dict={}\n",
    "    \n",
    "    for la in  old_labels:\n",
    "        # The below line tested if the dash character was causing labels to be misclassified.\n",
    "        li=la.replace(',','%2')\n",
    "        \n",
    "        label_x = li.upper()\n",
    "        if li[-1]=='p':\n",
    "            #   Note: this is for transcription factors (TF) from yeastract, which end in letter p.\n",
    "            upper_li=li[:-1]\n",
    "            label_x = upper_li.upper()\n",
    "            \n",
    "        if label_x in label_dict.keys() :\n",
    "            label_subset.append( li )\n",
    "            new_labels_dict[li] = label_dict[ label_x]\n",
    "        \n",
    "        elif label_x in full_label_list :\n",
    "            label_subset.append( li )\n",
    "            new_labels_dict[li] = label_x\n",
    "            \n",
    "        else:\n",
    "            #   print output for log.\n",
    "            print(' This label is incompatible:', li)            \n",
    "    return label_subset, new_labels_dict\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "old_cols, n= 3912\n",
      "Index(['PAU8', 'SEO1', 'YAL064W-B', 'TDA8', 'YAL064W', 'FLO9', 'NHP6B',\n",
      "       'YBR090C', 'TIM12', 'PHO3',\n",
      "       ...\n",
      "       'YBR012W-B', 'YDR210C-D', 'YGR161W-C', 'YHR214C-B', 'YPR158W-B',\n",
      "       'MAL61', 'MAL62', 'Uncharacterized.167', 'YDR524W-C', 'MAL63'],\n",
      "      dtype='object', length=3912)\n",
      "old_rows, n= 152\n",
      "['Smp1p' 'Ert1p' 'Thi2p' 'Mal33p' 'Kar4p' 'Hcm1p' 'Tup1p' 'Rds1p' 'Uga3p'\n",
      " 'Pho2p' 'Mbp1p' 'Rpn4p' 'Nrg1p' 'Pdc2p' 'Ino2p' 'Swi5p' 'Ume6p' 'Upc2p'\n",
      " 'Adr1p' 'Met32p' 'Yap6p' 'Sum1p' 'Aro80p' 'Cad1p' 'Yhp1p' 'Stp1p' 'Urc2p'\n",
      " 'Gcn4p' 'Mig3p' 'Gln3p' 'Aca1p' 'Flo8p' 'Swi4p' 'Rph1p' 'Tog1p' 'Hac1p'\n",
      " 'Gat1p' 'Pho4p' 'Fzf1p' 'Rtg2p' 'Hap2p' 'Mig2p' 'Cup2p' 'Tos8p' 'Hsf1p'\n",
      " 'Aft1p' 'Mig1p' 'Pdr1p' 'Rme1p' 'Mga1p' 'Rim101p' 'Opi1p' 'Yap3p' 'Stp2p'\n",
      " 'Ste12p' 'Ndt80p' 'Stb5p' 'Skn7p' 'Fkh1p' 'Xbp1p' 'Cst6p' 'Met28p'\n",
      " 'Yap5p' 'Dal81p' 'Mga2p' 'YJL206C' 'Gzf3p' 'Cbf1p' 'Ime1p' 'Rsf2p'\n",
      " 'Hms2p' 'Ash1p' 'Abf1p' 'Hap4p' 'Msn4p' 'Phd1p' 'Rgt1p' 'Ixr1p' 'Spt23p'\n",
      " 'Put3p' 'Dal80p' 'Bas1p' 'Gat3p' 'Ppr1p' 'Cha4p' 'Ace2p' 'Rfx1p' 'Ifh1p'\n",
      " 'Ecm22p' 'Hap1p' 'Sfp1p' 'Leu3p' 'Arg81p' 'Gal80p' 'Yox1p' 'Yap1p'\n",
      " 'Sok2p' 'Stb4p' 'Mac1p' 'Msn2p' 'Arg80p' 'Mcm1p' 'Mot3p' 'Mss11p' 'Hot1p'\n",
      " 'Rgm1p' 'Cat8p' 'Dal82p' 'Rap1p' 'Gcr2p' 'Sko1p' 'Met4p' 'Fkh2p' 'Crz1p'\n",
      " 'Msn1p' 'Ino4p' 'Hal9p' 'Rtg1p' 'Yap7p' 'Cin5p' 'Hms1p' 'Azf1p' 'Sfl1p'\n",
      " 'Yrr1p' 'Wtm1p' 'Tye7p' 'Hap5p' 'Pip2p' 'Ndd1p' 'Rdr1p' 'Gal4p' 'Usv1p'\n",
      " 'Aft2p' 'Cup9p' 'Rds2p' 'Tbf1p' 'Rlm1p' 'Gcr1p' 'Met31p' 'Haa1p' 'Rox1p'\n",
      " 'Fhl1p' 'Arr1p' 'Oaf1p' 'Rtg3p' 'Hap3p' 'Pdr3p' 'Reb1p' 'Tec1p' 'Sip4p'\n",
      " 'Zap1p' 'Mal63p']\n",
      "#  parsing columns:\n",
      " This label is incompatible: Uncharacterized\n",
      " This label is incompatible: SND3\n",
      " This label is incompatible: Uncharacterized.1\n",
      " This label is incompatible: Uncharacterized.2\n",
      " This label is incompatible: Uncharacterized.3\n",
      " This label is incompatible: ATG42\n",
      " This label is incompatible: Uncharacterized.4\n",
      " This label is incompatible: Uncharacterized.5\n",
      " This label is incompatible: DUR1%22\n",
      " This label is incompatible: Uncharacterized.6\n",
      " This label is incompatible: Uncharacterized.7\n",
      " This label is incompatible: Uncharacterized.8\n",
      " This label is incompatible: Uncharacterized.9\n",
      " This label is incompatible: Uncharacterized.10\n",
      " This label is incompatible: ELO2\n",
      " This label is incompatible: Uncharacterized.11\n",
      " This label is incompatible: Uncharacterized.12\n",
      " This label is incompatible: Uncharacterized.13\n",
      " This label is incompatible: Uncharacterized.14\n",
      " This label is incompatible: Uncharacterized.15\n",
      " This label is incompatible: EXP1\n",
      " This label is incompatible: Uncharacterized.16\n",
      " This label is incompatible: Uncharacterized.17\n",
      " This label is incompatible: Uncharacterized.18\n",
      " This label is incompatible: Uncharacterized.19\n",
      " This label is incompatible: Uncharacterized.20\n",
      " This label is incompatible: ATG31\n",
      " This label is incompatible: ILT1\n",
      " This label is incompatible: Uncharacterized.21\n",
      " This label is incompatible: Uncharacterized.22\n",
      " This label is incompatible: Uncharacterized.23\n",
      " This label is incompatible: Uncharacterized.24\n",
      " This label is incompatible: CCW22\n",
      " This label is incompatible: Uncharacterized.25\n",
      " This label is incompatible: Uncharacterized.26\n",
      " This label is incompatible: Uncharacterized.27\n",
      " This label is incompatible: MGP12\n",
      " This label is incompatible: YFT2\n",
      " This label is incompatible: Uncharacterized.28\n",
      " This label is incompatible: COI1\n",
      " This label is incompatible: Uncharacterized.29\n",
      " This label is incompatible: Uncharacterized.30\n",
      " This label is incompatible: Uncharacterized.31\n",
      " This label is incompatible: Uncharacterized.32\n",
      " This label is incompatible: Uncharacterized.33\n",
      " This label is incompatible: Uncharacterized.34\n",
      " This label is incompatible: Uncharacterized.35\n",
      " This label is incompatible: Uncharacterized.36\n",
      " This label is incompatible: Uncharacterized.37\n",
      " This label is incompatible: Uncharacterized.38\n",
      " This label is incompatible: Uncharacterized.39\n",
      " This label is incompatible: FRD1\n",
      " This label is incompatible: Uncharacterized.40\n",
      " This label is incompatible: GPP2\n",
      " This label is incompatible: ARG5%26\n",
      " This label is incompatible: Uncharacterized.41\n",
      " This label is incompatible: Uncharacterized.42\n",
      " This label is incompatible: FLO8\n",
      " This label is incompatible: Uncharacterized.43\n",
      " This label is incompatible: AAD16\n",
      " This label is incompatible: ZNF1\n",
      " This label is incompatible: LAM5\n",
      " This label is incompatible: Uncharacterized.44\n",
      " This label is incompatible: ADE5%27\n",
      " This label is incompatible: Uncharacterized.45\n",
      " This label is incompatible: Uncharacterized.46\n",
      " This label is incompatible: Uncharacterized.47\n",
      " This label is incompatible: YGK1\n",
      " This label is incompatible: Uncharacterized.48\n",
      " This label is incompatible: Uncharacterized.49\n",
      " This label is incompatible: Uncharacterized.50\n",
      " This label is incompatible: Uncharacterized.51\n",
      " This label is incompatible: EAT1\n",
      " This label is incompatible: DPC29\n",
      " This label is incompatible: Uncharacterized.52\n",
      " This label is incompatible: Uncharacterized.53\n",
      " This label is incompatible: Uncharacterized.54\n",
      " This label is incompatible: Uncharacterized.55\n",
      " This label is incompatible: PEX35\n",
      " This label is incompatible: Uncharacterized.56\n",
      " This label is incompatible: Uncharacterized.57\n",
      " This label is incompatible: MIC26\n",
      " This label is incompatible: Uncharacterized.58\n",
      " This label is incompatible: Uncharacterized.59\n",
      " This label is incompatible: Uncharacterized.60\n",
      " This label is incompatible: Uncharacterized.61\n",
      " This label is incompatible: Uncharacterized.62\n",
      " This label is incompatible: PAL2\n",
      " This label is incompatible: Uncharacterized.63\n",
      " This label is incompatible: ROF1\n",
      " This label is incompatible: Uncharacterized.64\n",
      " This label is incompatible: Uncharacterized.65\n",
      " This label is incompatible: CSS1\n",
      " This label is incompatible: YIL168W\n",
      " This label is incompatible: SDL1\n",
      " This label is incompatible: SOA1\n",
      " This label is incompatible: NDC80\n",
      " This label is incompatible: Uncharacterized.66\n",
      " This label is incompatible: Uncharacterized.67\n",
      " This label is incompatible: RCI37\n",
      " This label is incompatible: GPP1\n",
      " This label is incompatible: Uncharacterized.68\n",
      " This label is incompatible: EMC5\n",
      " This label is incompatible: VLD1\n",
      " This label is incompatible: FLO11\n",
      " This label is incompatible: NRE1\n",
      " This label is incompatible: Uncharacterized.69\n",
      " This label is incompatible: ROQ1\n",
      " This label is incompatible: Uncharacterized.70\n",
      " This label is incompatible: Uncharacterized.71\n",
      " This label is incompatible: TPH3\n",
      " This label is incompatible: MNN14\n",
      " This label is incompatible: Uncharacterized.72\n",
      " This label is incompatible: Uncharacterized.73\n",
      " This label is incompatible: Uncharacterized.74\n",
      " This label is incompatible: Uncharacterized.75\n",
      " This label is incompatible: Uncharacterized.76\n",
      " This label is incompatible: Uncharacterized.77\n",
      " This label is incompatible: APE1\n",
      " This label is incompatible: PSG1\n",
      " This label is incompatible: Uncharacterized.78\n",
      " This label is incompatible: Uncharacterized.79\n",
      " This label is incompatible: Uncharacterized.80\n",
      " This label is incompatible: Uncharacterized.81\n",
      " This label is incompatible: Uncharacterized.82\n",
      " This label is incompatible: PLN1\n",
      " This label is incompatible: Uncharacterized.83\n",
      " This label is incompatible: Uncharacterized.84\n",
      " This label is incompatible: Uncharacterized.85\n",
      " This label is incompatible: Uncharacterized.86\n",
      " This label is incompatible: MLO50\n",
      " This label is incompatible: EMA19\n",
      " This label is incompatible: Uncharacterized.87\n",
      " This label is incompatible: Uncharacterized.88\n",
      " This label is incompatible: Uncharacterized.89\n",
      " This label is incompatible: Uncharacterized.90\n",
      " This label is incompatible: Uncharacterized.91\n",
      " This label is incompatible: Uncharacterized.92\n",
      " This label is incompatible: Uncharacterized.93\n",
      " This label is incompatible: RSO55\n",
      " This label is incompatible: Uncharacterized.94\n",
      " This label is incompatible: ATG39\n",
      " This label is incompatible: Uncharacterized.95\n",
      " This label is incompatible: Uncharacterized.96\n",
      " This label is incompatible: LUG1\n",
      " This label is incompatible: ELO3\n",
      " This label is incompatible: Uncharacterized.97\n",
      " This label is incompatible: Uncharacterized.98\n",
      " This label is incompatible: Uncharacterized.99\n",
      " This label is incompatible: SEI1\n",
      " This label is incompatible: Uncharacterized.100\n",
      " This label is incompatible: Uncharacterized.101\n",
      " This label is incompatible: Uncharacterized.102\n",
      " This label is incompatible: Uncharacterized.103\n",
      " This label is incompatible: Uncharacterized.104\n",
      " This label is incompatible: Uncharacterized.105\n",
      " This label is incompatible: Uncharacterized.106\n",
      " This label is incompatible: Uncharacterized.107\n",
      " This label is incompatible: Uncharacterized.108\n",
      " This label is incompatible: NCW1\n",
      " This label is incompatible: DPI35\n",
      " This label is incompatible: Uncharacterized.109\n",
      " This label is incompatible: Uncharacterized.110\n",
      " This label is incompatible: Uncharacterized.111\n",
      " This label is incompatible: FPY1\n",
      " This label is incompatible: Uncharacterized.112\n",
      " This label is incompatible: Uncharacterized.113\n",
      " This label is incompatible: MLO1\n",
      " This label is incompatible: Uncharacterized.114\n",
      " This label is incompatible: Uncharacterized.115\n",
      " This label is incompatible: Uncharacterized.116\n",
      " This label is incompatible: Uncharacterized.117\n",
      " This label is incompatible: Uncharacterized.118\n",
      " This label is incompatible: Uncharacterized.119\n",
      " This label is incompatible: Uncharacterized.120\n",
      " This label is incompatible: PPN2\n",
      " This label is incompatible: Uncharacterized.121\n",
      " This label is incompatible: Uncharacterized.122\n",
      " This label is incompatible: MIC27\n",
      " This label is incompatible: Uncharacterized.123\n",
      " This label is incompatible: Uncharacterized.124\n",
      " This label is incompatible: Uncharacterized.125\n",
      " This label is incompatible: PYP1\n",
      " This label is incompatible: Uncharacterized.126\n",
      " This label is incompatible: PUL4\n",
      " This label is incompatible: Uncharacterized.127\n",
      " This label is incompatible: YOL153C\n",
      " This label is incompatible: Uncharacterized.128\n",
      " This label is incompatible: Uncharacterized.129\n",
      " This label is incompatible: Uncharacterized.130\n",
      " This label is incompatible: CRS5\n",
      " This label is incompatible: Uncharacterized.131\n",
      " This label is incompatible: TMC1\n",
      " This label is incompatible: Uncharacterized.132\n",
      " This label is incompatible: Uncharacterized.133\n",
      " This label is incompatible: Uncharacterized.134\n",
      " This label is incompatible: Uncharacterized.135\n",
      " This label is incompatible: Uncharacterized.136\n",
      " This label is incompatible: Uncharacterized.137\n",
      " This label is incompatible: Uncharacterized.138\n",
      " This label is incompatible: Uncharacterized.139\n",
      " This label is incompatible: Uncharacterized.140\n",
      " This label is incompatible: Uncharacterized.141\n",
      " This label is incompatible: Uncharacterized.142\n",
      " This label is incompatible: Uncharacterized.143\n",
      " This label is incompatible: Uncharacterized.144\n",
      " This label is incompatible: HTC1\n",
      " This label is incompatible: Uncharacterized.145\n",
      " This label is incompatible: Uncharacterized.146\n",
      " This label is incompatible: Uncharacterized.147\n",
      " This label is incompatible: Uncharacterized.148\n",
      " This label is incompatible: Uncharacterized.149\n",
      " This label is incompatible: Uncharacterized.150\n",
      " This label is incompatible: Uncharacterized.151\n",
      " This label is incompatible: Uncharacterized.152\n",
      " This label is incompatible: Uncharacterized.153\n",
      " This label is incompatible: Uncharacterized.154\n",
      " This label is incompatible: Uncharacterized.155\n",
      " This label is incompatible: YAR062W\n",
      " This label is incompatible: Uncharacterized.156\n",
      " This label is incompatible: Uncharacterized.157\n",
      " This label is incompatible: Uncharacterized.158\n",
      " This label is incompatible: Uncharacterized.159\n",
      " This label is incompatible: Uncharacterized.160\n",
      " This label is incompatible: Uncharacterized.161\n",
      " This label is incompatible: Uncharacterized.162\n",
      " This label is incompatible: IAI11\n",
      " This label is incompatible: Uncharacterized.163\n",
      " This label is incompatible: Uncharacterized.164\n",
      " This label is incompatible: Uncharacterized.165\n",
      " This label is incompatible: Uncharacterized.166\n",
      " This label is incompatible: LOG1\n",
      " This label is incompatible: COQ21\n",
      " This label is incompatible: IMP2'\n",
      " This label is incompatible: SNR190\n",
      " This label is incompatible: DPI8\n",
      " This label is incompatible: STA1\n",
      " This label is incompatible: MAL61\n",
      " This label is incompatible: MAL62\n",
      " This label is incompatible: Uncharacterized.167\n",
      " This label is incompatible: MAL63\n",
      "#  parsing rows:\n",
      " This label is incompatible: Flo8p\n",
      " This label is incompatible: Mal63p\n",
      "  number of compatible columns 3671\n",
      "  number of compatible rows 150\n"
     ]
    }
   ],
   "source": [
    "compatible_yeastract  =  relabel_matrix(  yeastract, gene_name_dict, full_gene_list )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>YAL068C</th>\n",
       "      <th>YAL067C</th>\n",
       "      <th>YAL064W-B</th>\n",
       "      <th>YAL064C-A</th>\n",
       "      <th>YAL064W</th>\n",
       "      <th>YAL063C</th>\n",
       "      <th>YBR089C-A</th>\n",
       "      <th>YBR090C</th>\n",
       "      <th>YBR091C</th>\n",
       "      <th>YBR092C</th>\n",
       "      <th>...</th>\n",
       "      <th>YPR036W-A</th>\n",
       "      <th>YPR158W-A</th>\n",
       "      <th>YBR012W-A</th>\n",
       "      <th>YHR047C</th>\n",
       "      <th>YBR012W-B</th>\n",
       "      <th>YDR210C-D</th>\n",
       "      <th>YGR161W-C</th>\n",
       "      <th>YHR214C-B</th>\n",
       "      <th>YPR158W-B</th>\n",
       "      <th>YDR524W-C</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3912</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YBR182C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR239C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR240C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR297W</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YCL055W</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBL005W</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR049C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR083W</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL089W</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJL056C</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 3671 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         YAL068C  YAL067C  YAL064W-B  YAL064C-A  YAL064W  YAL063C  YBR089C-A  \\\n",
       "3912                                                                           \n",
       "YBR182C        0        0          0          0        0        0          0   \n",
       "YBR239C        0        0          0          0        0        0          0   \n",
       "YBR240C        0        0          0          0        0        0          0   \n",
       "YBR297W        0        0          0          0        0        0          0   \n",
       "YCL055W        0        0          0          0        0        0          0   \n",
       "...          ...      ...        ...        ...      ...      ...        ...   \n",
       "YBL005W        0        0          0          0        0        0          0   \n",
       "YBR049C        0        0          0          0        0        0          0   \n",
       "YBR083W        0        0          0          0        0        0          0   \n",
       "YJL089W        0        0          0          0        0        0          0   \n",
       "YJL056C        0        0          0          0        0        0          0   \n",
       "\n",
       "         YBR090C  YBR091C  YBR092C  ...  YPR036W-A  YPR158W-A  YBR012W-A  \\\n",
       "3912                                ...                                    \n",
       "YBR182C        0        0        0  ...          0          0          0   \n",
       "YBR239C        0        0        0  ...          0          0          0   \n",
       "YBR240C        0        0        0  ...          0          0          0   \n",
       "YBR297W        0        0        0  ...          0          0          0   \n",
       "YCL055W        0        0        0  ...          0          0          0   \n",
       "...          ...      ...      ...  ...        ...        ...        ...   \n",
       "YBL005W        0        0        0  ...          0          0          0   \n",
       "YBR049C        0        0        0  ...          0          0          0   \n",
       "YBR083W        0        0        0  ...          0          0          1   \n",
       "YJL089W        0        0        0  ...          0          0          0   \n",
       "YJL056C        0        0        0  ...          0          0          0   \n",
       "\n",
       "         YHR047C  YBR012W-B  YDR210C-D  YGR161W-C  YHR214C-B  YPR158W-B  \\\n",
       "3912                                                                      \n",
       "YBR182C        0          0          0          0          0          0   \n",
       "YBR239C        0          0          0          0          0          0   \n",
       "YBR240C        0          0          0          0          0          0   \n",
       "YBR297W        0          0          0          0          0          0   \n",
       "YCL055W        0          0          0          0          0          0   \n",
       "...          ...        ...        ...        ...        ...        ...   \n",
       "YBL005W        0          0          0          0          0          0   \n",
       "YBR049C        0          0          0          0          0          0   \n",
       "YBR083W        0          1          0          0          0          1   \n",
       "YJL089W        0          0          0          0          0          0   \n",
       "YJL056C        0          0          0          0          0          0   \n",
       "\n",
       "         YDR524W-C  \n",
       "3912                \n",
       "YBR182C          0  \n",
       "YBR239C          0  \n",
       "YBR240C          0  \n",
       "YBR297W          0  \n",
       "YCL055W          0  \n",
       "...            ...  \n",
       "YBL005W          0  \n",
       "YBR049C          0  \n",
       "YBR083W          0  \n",
       "YJL089W          0  \n",
       "YJL056C          0  \n",
       "\n",
       "[150 rows x 3671 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compatible_yeastract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "### remove columns whose names are not there in the expression matrix ######### \n",
    "expression_genes=pan.read_csv('/data/users/cs20s037/CITNonLinear/findr/findrfiles/scripts/list_genes_in_order_01.csv',header=None)[0].to_list()\n",
    "commoncolumns= [item for item in expression_genes if item in compatible_yeastract.columns]\n",
    "commonrows=[item for item in expression_genes if item in compatible_yeastract.index]\n",
    "compatible_yeastract= compatible_yeastract.loc[commonrows,commoncolumns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common Rows 148\n",
      "Common columns 3422\n"
     ]
    }
   ],
   "source": [
    "print(\"Common Rows\",len(commonrows))\n",
    "print(\"Common columns\",len(commoncolumns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pan.read_csv('/data/users/cs20s037/CITNonLinear/findr/findrfiles/scripts/strongest_eqtls_r_3columns.csv')\n",
    "df.drop(columns=df.columns[0], axis=1,  inplace=True)\n",
    "df.columns=['cis','pmarker','r']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cis</th>\n",
       "      <th>pmarker</th>\n",
       "      <th>r</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YAL003W</td>\n",
       "      <td>chrI:133174_G/A</td>\n",
       "      <td>0.383781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YAL008W</td>\n",
       "      <td>chrI:136961_T/C</td>\n",
       "      <td>0.273774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YAL009W</td>\n",
       "      <td>chrI:132723_G/A</td>\n",
       "      <td>-0.260297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YAL010C</td>\n",
       "      <td>chrI:134219_C/T</td>\n",
       "      <td>0.196381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YAL011W</td>\n",
       "      <td>chrI:131539_C/T</td>\n",
       "      <td>-0.302313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2879</th>\n",
       "      <td>YPR180W</td>\n",
       "      <td>chrXVI:897736_G/A</td>\n",
       "      <td>0.229310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2880</th>\n",
       "      <td>YPR186C</td>\n",
       "      <td>chrXVI:907815_T/C</td>\n",
       "      <td>0.131461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2881</th>\n",
       "      <td>YPR191W</td>\n",
       "      <td>chrXVI:921485_T/C</td>\n",
       "      <td>0.146219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882</th>\n",
       "      <td>YPR192W</td>\n",
       "      <td>chrXVI:922297_T/C</td>\n",
       "      <td>0.790686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2883</th>\n",
       "      <td>YPR193C</td>\n",
       "      <td>chrXVI:922883_C/CCTA</td>\n",
       "      <td>0.713753</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2884 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          cis               pmarker         r\n",
       "0     YAL003W       chrI:133174_G/A  0.383781\n",
       "1     YAL008W       chrI:136961_T/C  0.273774\n",
       "2     YAL009W       chrI:132723_G/A -0.260297\n",
       "3     YAL010C       chrI:134219_C/T  0.196381\n",
       "4     YAL011W       chrI:131539_C/T -0.302313\n",
       "...       ...                   ...       ...\n",
       "2879  YPR180W     chrXVI:897736_G/A  0.229310\n",
       "2880  YPR186C     chrXVI:907815_T/C  0.131461\n",
       "2881  YPR191W     chrXVI:921485_T/C  0.146219\n",
       "2882  YPR192W     chrXVI:922297_T/C  0.790686\n",
       "2883  YPR193C  chrXVI:922883_C/CCTA  0.713753\n",
       "\n",
       "[2884 rows x 3 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtlgenes=df['cis']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "yeastractgenes=compatible_yeastract.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "123"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(yeastractgenes)\n",
    "len(set(eqtlgenes).intersection(set(yeastractgenes)))  ## the number of trios that we can take is 80*3671 # 80 of the TFs are cis eQTLs\n",
    "len(set(eqtlgenes).intersection(set(compatible_yeastract.columns))) ## 1710 out of the total 3671 columns genes are cis eQTLs \n",
    "len(set(yeastractgenes).intersection(set(compatible_yeastract.columns))) ## 125 TFs are both rows and columns, should we remove self regulation ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#common tfs present in both yeastract and ciseqtl file\n",
    "commontfs = [item for item in eqtlgenes if item in yeastractgenes] # not using set, because reproducing the result is not possible\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#self regulations not removed, remove them while processing ? \n",
    "#cit gives NA values if both A and B are the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save common tfs\n",
    "np.savetxt(\"commontfs.csv\",\n",
    "        commontfs,\n",
    "        delimiter =\", \",\n",
    "        fmt ='% s')\n",
    "#commontfs_test=pan.read_csv('commontfs.csv',header=None)[0].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#take only those rows that are common \n",
    "subset_yeast=compatible_yeastract.loc[commontfs,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######### Do a wilcoxin test unpaired #############"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "subsetrownames=subset_yeast.index\n",
    "subsetcolnames=subset_yeast.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "### create cis genotype dictionary ##### \n",
    "cisgeno={}\n",
    "for i in range(0,len(df)):\n",
    "    cisgeno[df['cis'][i]]=df['pmarker'][i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the expression and genotype data \n",
    "expression_data=pan.read_csv(\"/data/users/cs20s037/CITNonLinear/findr/findrfiles/scripts/expression_reordered_01.csv\",index_col=0,header=0)\n",
    "genotype=pan.read_csv(\"/data/users/cs20s037/CITNonLinear/findr/findrfiles/scripts/genotypes_binary_strongest_eqtl.csv\",header=0,index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##set seed## \n",
    "#seed generated using the following command \n",
    "# dont run it now, already used \n",
    "#import random\n",
    "#import time\n",
    "# Generate a seed using current time\n",
    "#seed = int(time.time()) \n",
    "# seed = 1688557140 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set the seed and dont set the seed while calling the sample function, it will reproduce the same results\n",
    "np.random.seed( 1688557140 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "causal=[]\n",
    "indep=[]\n",
    "for j in range(0,len(subset_yeast)):\n",
    "    # go through each row \n",
    "    groundtruth=subset_yeast.iloc[j,]\n",
    "    Aname=subsetrownames[j]\n",
    "    indice_list=[]\n",
    "    L=genotype.loc[:,cisgeno[Aname]]\n",
    "    #going to take only those who pass the wilkoxin test\n",
    "    unique_values = np.unique(L)\n",
    "    for value in unique_values:\n",
    "        indices = np.where(L == value)[0]\n",
    "        indice_list.append(indices)\n",
    "\n",
    "    pvals=[]\n",
    "    cols_used=[]\n",
    "    for i in range(0,len(subsetcolnames)):\n",
    "        Bname=subsetcolnames[i]\n",
    "        if Aname==Bname: # remove self regulations\n",
    "            continue\n",
    "        cols_used.append(i)\n",
    "        Bexpression=expression_data.loc[:,Bname]\n",
    "    \n",
    "        datasplit=[]\n",
    "    \n",
    "        datasplit.append(Bexpression[indice_list[0]])\n",
    "        datasplit.append(Bexpression[indice_list[1]])\n",
    "        pvals.append(ranksums(datasplit[0],datasplit[1])[1])\n",
    "    pvals=multipletests(pvals, alpha=0.05, method='fdr_bh')[1] \n",
    "    comparison = [val < 0.2 for val in pvals]\n",
    "    comp=pan.DataFrame({'colnames':subsetcolnames[cols_used].to_list(),'groundtruth':groundtruth[cols_used].to_list(),'pvalue':comparison})\n",
    "    gt1=comp[(comp['groundtruth']==1) & (comp['pvalue'] == True)]\n",
    "    gt1.insert(0,'A',Aname)\n",
    "    gt1.insert(0,'L',cisgeno[Aname])\n",
    "    causal.extend(gt1[['L','A','colnames','groundtruth']].values.tolist())\n",
    "    # sample equal number of inpendent trios\n",
    "    #seed set \n",
    "    assert len(gt1)<=len(comp[(comp['groundtruth']==0) & (comp['pvalue'] == True)])\n",
    "    gt2=comp[(comp['groundtruth']==0) & (comp['pvalue'] == True)].sample(n=len(gt1))   #,random_state=1688557140)\n",
    "    gt2.insert(0,'A',Aname)\n",
    "    gt2.insert(0,'L',cisgeno[Aname])\n",
    "    indep.extend(gt2[['L','A','colnames','groundtruth']].values.tolist())\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "causal_wilko=pan.DataFrame(causal,columns=['pmarker','cis','targetgene','groundtruth'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1234"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(causal_wilko)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "indp_wilko=pan.DataFrame(indep,columns=['pmarker','cis','targetgene','groundtruth'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1234"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(indp_wilko)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "## for yeast all causal data ########\n",
    "#1257 was with self-regulations \n",
    "#1234 is without self-regulations\n",
    "f = open(\"yeastgt_1_wilko1234_ready.txt\", \"w\")\n",
    "for i in range(len(causal_wilko)):\n",
    "    sample=causal_wilko.iloc[i,:]\n",
    "    L=genotype.loc[:,sample['pmarker']]\n",
    "    A=expression_data.loc[:,sample['cis']]\n",
    "    B=expression_data.loc[:,sample['targetgene']]\n",
    "    trioheader= sample['pmarker']+' '+sample['cis']+' '+sample['targetgene']\n",
    "    f.write(trioheader)\n",
    "    f.write(\"\\n\")\n",
    "    np.savetxt(f,L,newline=' ',fmt='%s')\n",
    "    f.write(\"\\n\")\n",
    "    np.savetxt(f,A,newline=' ',fmt='%s')\n",
    "    f.write(\"\\n\")\n",
    "    np.savetxt(f,B,newline=' ',fmt='%s')\n",
    "    f.write(\"\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "## for yeast all independent data ########\n",
    "f = open(\"yeastgt_0_wilko1234_ready.txt\", \"w\")\n",
    "for i in range(len(indp_wilko)):\n",
    "    sample=indp_wilko.iloc[i,:]\n",
    "    L=genotype.loc[:,sample['pmarker']]\n",
    "    A=expression_data.loc[:,sample['cis']]\n",
    "    B=expression_data.loc[:,sample['targetgene']]\n",
    "    trioheader= sample['pmarker']+' '+sample['cis']+' '+sample['targetgene']\n",
    "    f.write(trioheader)\n",
    "    f.write(\"\\n\")\n",
    "    np.savetxt(f,L,newline=' ',fmt='%s')\n",
    "    f.write(\"\\n\")\n",
    "    np.savetxt(f,A,newline=' ',fmt='%s')\n",
    "    f.write(\"\\n\")\n",
    "    np.savetxt(f,B,newline=' ',fmt='%s')\n",
    "    f.write(\"\\n\")\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
